### PR TITLE
fixes #3930 - refactoring of edit_self implementation.

### PR DESCRIPTION
**Pull request by ohadlevy**

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -43,11 +42,8 @@ def edit
 
   def update
     # Remove keys for restricted variables when the user is editing their own account
-    if editing_self
-      for key in params[:user].keys
-        params[:user].delete key unless %w{password_confirmation password mail firstname lastname locale}.include? key
-      end
-      User.current.editing_self = true
+    if params[:users] && editing_self?
```
Comment (abenari) [inline app/controllers/users_controller.rb]: checking params[:users] and using params[:user] could be a typo

File: app/models/user.rb [hunk 2]
```diff
@@ -190,7 +190,11 @@ def manage_password?
   # * a permission Symbol (eg. :edit_project)
   def allowed_to?(action, options={})
```
Comment (abenari) [inline app/models/user.rb]: can we remove options here? we are not using them in the method

File: test/functional/users_controller_test.rb [hunk 3]
```diff
@@ -220,4 +221,28 @@ def test_one #"should not remove the anonymous role" do
     assert_redirected_to edit_user_path(User.find_by_login('ares'))
   end
 
+  test 'non admin user should edit itself' do
+    User.current = users(:one)
+    get :edit, { :id => User.current.id }
+    assert_response :success
```
Comment (abenari) [inline test/functional/users_controller_test.rb:227]: assert_response :redirect

File: test/functional/users_controller_test.rb [hunk 4]
```diff
@@ -220,4 +221,28 @@ def test_one #"should not remove the anonymous role" do
...
+  test 'non admin user should be able to update itself' do
+    User.current = users(:one)
+    put :update, { :id => users(:one).id, :user => { :firstname => 'test' } }
+    assert_response :success
+  end
+
+  test 'non admin user should not be able to edit another user' do
+    User.current = users(:one)
+    get :edit, { :id => users(:two) }
+    assert_equal @response.status, 403
```
Comment (abenari) [inline test/functional/users_controller_test.rb]: nitpick assert_response 403 is shorter

File: test/functional/users_controller_test.rb [hunk 5]
```diff
@@ -220,4 +221,28 @@ def test_one #"should not remove the anonymous role" do
...
+  end
+
+  test 'non admin user should not be able to update another user' do
+    User.current = users(:one)
+    put :update, { :id => users(:two).id, :user => { :firstname => 'test' } }
+    assert_equal @response.status, 403
```
Comment (abenari) [inline test/functional/users_controller_test.rb]: same as above

**Comment by abenari** [comment]
do we want to refactor the api controller too?

**Comment by ohadlevy** [comment]
[test]

**Comment by ohadlevy** [comment]
@abenari please review?

**Comment by abenari** [comment]
merged as 355bce36288ec6cd9f5a66b656f8a84158f2a8e1

**Pull request closed without merging**